src/coredump/stacktrace.c

```diff
@@ -342,63 +367,37 @@ static int parse_core(int fd, const char *executable, char **ret, JsonVariant **
         elf_version(EV_CURRENT);
 
         c.elf = elf_begin(fd, ELF_C_READ_MMAP, NULL);
-        if (!c.elf) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (!c.elf)
+                return -EINVAL;
 
         c.dwfl = dwfl_begin(&callbacks);
-        if (!c.dwfl) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (!c.dwfl)
+                return -EINVAL;
 
-        if (dwfl_core_file_report(c.dwfl, c.elf, executable) < 0) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (dwfl_core_file_report(c.dwfl, c.elf, executable) < 0)
+                return -EINVAL;
 
-        if (dwfl_report_end(c.dwfl, NULL, NULL) != 0) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (dwfl_report_end(c.dwfl, NULL, NULL) != 0)
+                return -EINVAL;
 
-        if (dwfl_getmodules(c.dwfl, &module_callback, &c, 0) < 0) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (dwfl_getmodules(c.dwfl, &module_callback, &c, 0) < 0)
+                return -EINVAL;
 
-        if (dwfl_core_file_attach(c.dwfl, c.elf) < 0) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (dwfl_core_file_attach(c.dwfl, c.elf) < 0)
+                return -EINVAL;
 
-        if (dwfl_getthreads(c.dwfl, thread_callback, &c) < 0) {
-                r = -EINVAL;
-                goto finish;
-        }
+        if (dwfl_getthreads(c.dwfl, thread_callback, &c) < 0)
+                return -EINVAL;
 
         r = fflush_and_check(c.f);
         if (r < 0)
-                goto finish;
-        c.f = safe_fclose(c.f);
+                return r;
```

Comment:
Nope, sorry. `buf` may be invalidated when the file is closed.

Reply:
so we need to free before closing?

Reply:
the `_cleanup_free` comes after the `stack_context_destroyp` so the order should be correct?

Reply:
added a comment mentioning the order, as we do in other places

Reply:
No, this is not about the order of the cleanup functions. The problem is the pointer used by the memstream can be invalidated at various points when we're writing to the stream, or flushing the stream, or when closing the stream. So to get a pointer that is valid after the function exits, you must first close the stream, and only then do `TAKE_PTR(buf)`. This cannot be expressed with cleanup functions alone.

Reply:
ah, got it - so this should work, while keeping the `_cleanup_` in place, right?

```
if (ret) {
        c.f = safe_fclose(c.f);
       *ret = TAKE_PTR(buf);
}
```

Reply:
Yep.

Reply:
ok, done - thanks for the tip, I had completely forgot about all of that